service/src/main/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilder.java

```
@@ -0,0 +1,206 @@
+    policyMap.put(
+        ControlledResourceIamRole.EDITOR,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.EDITOR.toSamRole()));
+    return policyMap;
+  }
+
+  /**
+   * Given the map created in @{@link #makePolicyMap()}, populate the resource request with the
+   * policies.
+   *
+   * @param request resource request for setting the roles on the Sam resource creation
+   * @param policyMap previously created policy map
+   */
+  private void applyPolicyMap(
+      CreateResourceRequestV2 request,
+      Map<ControlledResourceIamRole, AccessPolicyMembershipV2> policyMap) {
+    for (Entry<ControlledResourceIamRole, AccessPolicyMembershipV2> entry : policyMap.entrySet()) {
+      request.putPoliciesItem(entry.getKey().toSamRole(), entry.getValue());
+    }
+  }
+
+  /**
+   * Add WSM's service account as the owner of a controlled resource in Sam. Used for admin
+   * reassignment of resources. This assumes samService.initialize() has already been called, which
+   * should happen on start.
+   */
+  private void addWsmResourceOwnerPolicy(CreateResourceRequestV2 request)
+      throws InterruptedException {
+    try {
+      AuthenticatedUserRequest wsmRequest =
+          new AuthenticatedUserRequest().token(Optional.of(samService.getWsmServiceAccountToken()));
+      String wsmSaEmail = samService.getUserEmailFromSam(wsmRequest);
+      AccessPolicyMembershipV2 ownerPolicy =
+          new AccessPolicyMembershipV2()
+              .addRolesItem(ControlledResourceIamRole.OWNER.toSamRole())
+              .addMemberEmailsItem(wsmSaEmail);
+      request.putPoliciesItem(ControlledResourceIamRole.OWNER.toSamRole(), ownerPolicy);
+    } catch (InternalServerErrorException e) {
+      // In cases where WSM is not running as a service account (e.g. unit tests), the above call to
+      // get application default credentials will fail. This is fine, as those cases don't create
+      // real resources.
+      logger.warn(
```

Was just thinking if this fails for other reasons in the real world, then we'd swallow this exception. Not sure what the risk/impact is though, so not a strong objection.

---

True. It would be very bad if this happened in production. I'll make a ticket to track.